File: solutions_2453486_1/C++/anttil/tictac.cpp
#include <iostream>

using namespace std;

int n;
string t[4];

bool voitto(int y, int x, int ys, int xs, char m) {
    for (int i = 0; i < 4; i++) {
        if (t[y][x] != m && t[y][x] != 'T') {
            return false;
        }
        y += ys;
        x += xs;
    }
    return true;
}

bool valmis() {
    for (int i = 0; i < 4; i++) {
        for (int j = 0; j < 4; j++) {
            if (t[i][j] == '.') return false;
        }
    }
    return true;
}

void tutki(int n) {
    char v = 'D';
    for (int i = 0; i < 4; i++) {
        if (voitto(0, i, 1, 0, 'X')) v = 'X';
        if (voitto(0, i, 1, 0, 'O')) v = 'O';
        if (voitto(i, 0, 0, 1, 'X')) v = 'X';
        if (voitto(i, 0, 0, 1, 'O')) v = 'O';
    }
    if (voitto(0, 0, 1, 1, 'X')) v = 'X';
    if (voitto(0, 0, 1, 1, 'O')) v = 'O';
    if (voitto(3, 0, -1, 1, 'X')) v = 'X';
    if (voitto(3, 0, -1, 1, 'O')) v = 'O';
    cout << "Case #" << n << ": ";
    if (v == 'D') {
        if (valmis()) {
            cout << "Draw\n";
        } else {
            cout << "Game has not completed\n";
        }
    } else {
        cout << v << " won\n";
    }
}

int main() {
    cin >> n;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < 4; j++) {
            cin >> t[j];
        }
        tutki(i+1);
    }
}

File: solutions_2453486_1/Python/smelendez/tictac.py
#!/usr/bin/python

import sys

def checkrow(row):
    winner = None
    if row.count('X') >=3 and row.count('.') == 0 and row.count('O') == 0:
        winner = 'X won'
    elif row.count('O') >=3 and row.count('.') == 0 and row.count('X') == 0:
        winner = 'O won'
    return winner

def main():
    infile = file(sys.argv[1])

    ncases = int(infile.readline().strip())

    for case in xrange(ncases):
        rows = []
        winner = None

        for j in xrange(4):
            rows.append(infile.readline().strip())

        for row in rows:
            if not winner:
                winner = checkrow(row)

        for col in xrange(4):
            column = [row[col] for row in rows]
            if not winner:
                winner = checkrow(column)

        diag = [rows[i][i] for i in xrange(4)]
        if not winner:
            winner = checkrow(diag)

        diag = [rows[i][3-i] for i in xrange(4)]
        if not winner:
            winner =checkrow(diag)

        if not winner:
            for row in rows:
                if '.' in row:
                    winner = "Game has not completed"

        if not winner: winner = "Draw" 
        
        print "Case #%d: %s" % (case+1, winner)
        infile.readline()

        
            






        

        
main()
